### PR TITLE
Allow to upgrade service offerings from local <> shared storage pools

### PR DESCRIPTION
**Pull request by GabrielBrascher**
### Description

This PR addresses the issue raised at #4545 (_Fail to change Service offering from local <> shared storage_).

When upgrading a VM service offering it is validated if the new offering has the same storage scope (local or shared) as the current offering. I think that the validation makes sense in a way of preventing running Root disks with an offering that does not match the current storage pool. However, the validation only compares both offerings and does not consider that it is possible to migrate Volumes between local <> shared storage pools.

The idea behind this implementation is that CloudStack should check the scope of the current storage pool which the ROOT volume is allocated; this, it is possible to migrate the volume between storage pools and list/upgrade according to the offerings that are supported for such pool.

This PR also fixes an issue where the API command that lists offerings for a VM should follow the same idea and list based on the storage pool that the volume is allocated and not the previous offering.

Fixes: #4545

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [X] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [X] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?
**Prior the fix:**
1. Migrate volume from shared to local (works)
2. update VM offering for local storage (fails)

**With the fix:**
1. Migrate volume from shared to local (works)
2. update VM offering for local storage (works if the offering was configured for local storage, if it was set for shared storage then it fails)

### CODE REVIEW

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@GabrielBrascher this PR has milestone 4.15.1.0 but base branch master, pl fix thanks.

**Comment by GabrielBrascher** [comment]
The branch has been rebased to 4.15 and PR is pointed to 4.15.
@rhtyd thanks for pointing it out.

Implementation based on original commits located at branch https://github.com/PCextreme/cloudstack/tree/migrate-local-shared-master

Commits squashed and rebased were:
- [57646ec1bb61e77b4e4343e8f4dc36739ec1aeb5](https://github.com/PCextreme/cloudstack/commit/57646ec1bb61e77b4e4343e8f4dc36739ec1aeb5)
- [6b848766e677cde3ae5bbaffee684b2c60e3593a](https://github.com/PCextreme/cloudstack/commit/6b848766e677cde3ae5bbaffee684b2c60e3593a)
- [9ae248314bbfbc026041425b2afae2197b30c8b6](https://github.com/PCextreme/cloudstack/commit/9ae248314bbfbc026041425b2afae2197b30c8b6)
- [cbf635829b8221f1645e916ea32c08e81aaf52e9](https://github.com/PCextreme/cloudstack/commit/cbf635829b8221f1645e916ea32c08e81aaf52e9)

**Comment by GabrielBrascher** [comment]
@blueorangutan package

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 1]
```diff
@@ -3706,6 +3691,34 @@ public void checkIfCanUpgrade(final VirtualMachine vmInstance, final ServiceOffe
...
     }
 
+    protected void checkIfNewOfferingStorageScopeMatchesStoragePool(VirtualMachine vmInstance, ServiceOffering newServiceOffering) {
+        boolean isRootVolumeOnLocalStorage = isRootVolumeOnLocalStorage(vmInstance.getId());
+
+        if (newServiceOffering.isUseLocalStorage() && !isRootVolumeOnLocalStorage) {
+            String message = String .format("Unable to upgrade virtual machine %s, target offering use local storage but the storage pool where "
+                    + "the volume is allocated is a shared storage.", vmInstance.toString());
+            throw new InvalidParameterValueException(message);
+        } else if (!newServiceOffering.isUseLocalStorage() && isRootVolumeOnLocalStorage) {
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:3701]: else is not needed here and might cause confusion
```suggestion
        } 
        if (!newServiceOffering.isUseLocalStorage() && isRootVolumeOnLocalStorage) {
```

**Review by DaanHoogland — APPROVED** [review]

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 1 above)
Comment (GabrielBrascher) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: Thanks for the suggestion @DaanHoogland, makes sense :+1:

**Comment by GabrielBrascher** [comment]
I update PR adding a commit addressing @DaanHoogland's proposed change as well as adding a Javadoc block to make it "dev-friendly" :-)

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by shwstppr** [comment]
@GabrielBrascher can you please resolve conflicts

**Comment by GabrielBrascher** [comment]
@shwstppr thanks for the heads up. Code has been updated fixing the conflict.

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Review by shwstppr — APPROVED** [review]
Code LGTM.
Changes correctly check storage and new offering as the old offering can give wrong details as migration between different storage types is allowed and there could be DB inconsistency wrt service offering.

**Review by Pearl1594 — APPROVED** [review]
Tested. LGTM

**Pull request merged**